lib/mongo/server/connectable.rb

```diff
@@ -88,10 +88,11 @@ def ensure_connected
         ensure_same_process!
         connect!
         begin
-          yield socket
-        rescue Exception => e
-          disconnect!
-          raise e
+          res = yield socket
+          success = true
+          res
+        ensure
+          success or disconnect!
```

Comment:
If an exception is raised when the socket is yielded, it won't return a value. If I'm not mistaken, it's never the case that result has a value *and* the block to which the socket is yielded throws an exception. What do you think?